master/internal/api_user.go

```diff
@@ -0,0 +1,151 @@
+			AgentGroup: user.AgentGroup.ValueOrZero(),
+		}
+	}
+	return &userv1.User{
+		Id:             int32(user.ID),
+		Username:       user.Username,
+		Admin:          user.Admin,
+		Active:         user.Active,
+		AgentUserGroup: agentUserGroup,
+	}
+}
+
+func getUser(d *db.PgDB, username string) (*userv1.User, error) {
+	user, err := d.UserByUsername(username)
+	switch {
+	case err == db.ErrNotFound:
+		return nil, errUserNotFound
+	case err != nil:
+		return nil, err
+	}
+	var protoAug *userv1.AgentUserGroup
+	agentUserGroup, err := d.AgentUserGroup(user.ID)
+	if agentUserGroup != nil {
+		protoAug = &userv1.AgentUserGroup{
+			AgentUid:   int32(agentUserGroup.UID),
+			AgentGid:   int32(agentUserGroup.GID),
+			AgentUser:  agentUserGroup.User,
+			AgentGroup: agentUserGroup.Group,
+		}
+	}
+	return &userv1.User{
+		Id:             int32(user.ID),
+		Username:       user.Username,
+		Admin:          user.Admin,
+		Active:         user.Active,
+		AgentUserGroup: protoAug,
+	}, err
+}
+
+func (a *apiServer) GetUsers(
+	context.Context, *apiv1.GetUsersRequest) (*apiv1.GetUsersResponse, error) {
+	users, err := a.m.db.UserList()
+	if err != nil {
+		return nil, err
+	}
+	result := &apiv1.GetUsersResponse{}
+	for _, user := range users {
+		result.Users = append(result.Users, toProtoUserFromFullUser(user))
+	}
+	sort.Slice(result.Users, func(i, j int) bool {
+		return result.Users[i].Username < result.Users[j].Username
+	})
+	return result, nil
+}
+
+func (a *apiServer) GetUser(
+	_ context.Context, req *apiv1.GetUserRequest) (*apiv1.GetUserResponse, error) {
+	fullUser, err := getUser(a.m.db, req.Username)
+	return &apiv1.GetUserResponse{User: fullUser}, err
+}
+
+func (a *apiServer) PostUser(
+	ctx context.Context, req *apiv1.PostUserRequest) (*apiv1.PostUserResponse, error) {
+	if err := grpc.ValidateRequest(
+		func() (bool, string) { return req.User != nil, "no user specified" },
+		func() (bool, string) { return req.User.Username != "", "no username specified" },
+	); err != nil {
+		return nil, err
+	}
+	curUser, _, err := grpc.GetUser(ctx, a.m.db)
+	if err != nil {
+		return nil, err
+	}
+	if !curUser.Admin {
+		return nil, grpc.ErrPermissionDenied
+	}
+	user := &model.User{
+		Username: req.User.Username,
+		Admin:    req.User.Admin,
+		Active:   req.User.Active,
+	}
+	if err = user.UpdatePasswordHash(req.Password); err != nil {
+		return nil, err
+	}
+	var agentUserGroup *model.AgentUserGroup
+	if req.User.AgentUserGroup != nil {
+		agentUserGroup = &model.AgentUserGroup{
+			User:  req.User.AgentUserGroup.AgentUser,
+			UID:   int(req.User.AgentUserGroup.AgentUid),
+			Group: req.User.AgentUserGroup.AgentGroup,
+			GID:   int(req.User.AgentUserGroup.AgentGid),
+		}
+	}
+
+	switch err = a.m.db.AddUser(user, agentUserGroup); {
+	case err == db.ErrDuplicateRecord:
+		return nil, status.Error(codes.InvalidArgument, "user already exists")
+	case err != nil:
+		return nil, err
+	}
+	fullUser, err := getUser(a.m.db, req.User.Username)
```

Comment:
question: seems odd to have to get the user back from the db after putting it in there, is this avoidable?

Reply:
I am planning on cleaning up the database queries and interface after this project is done. For now, I am keeping it as is.